### PR TITLE
Maintenance: update vendored minizip code to the version distributed with zlib v1.3

### LINKED ISSUE #240: Minizip
I see here:
- https://github.com/xroche/httrack/tree/master/src/minizip

It will be nice to have compatibility with current Minizip (included in zlib 1.2.13) and Minizip NG?
- https://github.com/madler/zlib/tree/master/contrib/minizip
- https://github.com/madler/zlib/tree/develop/contrib/minizip
- https://github.com/zlib-ng/minizip-ng

### PR DESCRIPTION
**Pull request by jayaddison**
Resolves #240.

After updating `minizip` to match `zlib` v1.3 and manually re-applying the `.diff` files, I ran the following checks:

* Compiled the code.
* Checked that the resulting `httrack` binary is runnable.
* Ran some basic mirroring checks on a website, including use of the `--store-all-in-cache` option, and confirmed that a valid `.zip` cache file was created.
* Ran the test suite from the `tests` directory (output: `all tests passed`).

Applying the `.diff` files was definitely the most difficult part of the process, and likely the most difficult to review, also.

### CODE REVIEW

**Comment by xroche** [comment]
Thank you for this fix! Reviewed, tested with both built-in tests and manually downloading.

````
PASS: 00_runnable.test
PASS: 01_engine-charset.test
PASS: 01_engine-entities.test
PASS: 01_engine-hashtable.test
PASS: 01_engine-idna.test
PASS: 01_engine-simplify.test
PASS: 10_crawl-simple.test
PASS: 11_crawl-cookies.test
PASS: 11_crawl-idna.test
PASS: 11_crawl-international.test
PASS: 11_crawl-longurl.test
PASS: 11_crawl-parsing.test
PASS: 12_crawl_https.test
````

I already cherry-picked https://github.com/madler/zlib/commit/73331a6a0481067628f065ffe87bb1d8f787d10c as advised, but updating this code seems a good idea!

Thanks again for your help,

**Comment by jayaddison** [comment]
You're welcome - thank you!

**Comment by Neustradamus** [comment]
@jayaddison: Thanks for your change but can you add the CVE-2023-45853 patch because it is not included in Zlib 1.3...
- https://github.com/madler/zlib/pull/843
- https://github.com/madler/zlib/issues/868

You can see the last link...

**Comment by Neustradamus** [comment]
@jayaddison: Thanks for your PR!

@xroche: Thanks for the CVE-2023-45853 patch because it is not included in Zlib 1.3...
- https://github.com/madler/zlib/pull/843
- https://github.com/madler/zlib/issues/868

File: src/minizip/zip.c.diff [hunk 1]
```diff
@@ -1,97 +1,15 @@
...
-@@ -1450,7 +1450,7 @@
+--- zip.c.orig	2023-11-17 16:21:50.363864777 +0000
++++ zip.c	2023-11-17 16:26:04.671858368 +0000
+@@ -1401,7 +1401,7 @@
      else
  #endif
      {
--      zi->ci.stream.next_in = (Bytef*)buf;
+-      zi->ci.stream.next_in = (Bytef*)(uintptr_t)buf;
 +      zi->ci.stream.next_in = buf;
```
Comment (jayaddison) [inline src/minizip/zip.c.diff:8]: @xroche viewing a diff-of-a-diff can be confusing, but I'd like to explain and highlight this change -- it was the only case where I was uncertain about how to resolve a merge conflict during the upgrade.

Basically: an additional typecast was added to the `minizip` sources here, and I've confirmed that that was intentional.

It caused this part of the diff/patch to not apply cleanly, and the resolution I chose for the pull request was to leave the assignment of `next_in = buf` as-is, without any typecasts.

However: I'm not totally confident about it.  It's a small detail, but I wanted to mention it anyway.

**Comment by jayaddison** [comment]
@xroche a note that [`zlib` v1.3.1](https://github.com/madler/zlib/releases/tag/v1.3.1) is now available including the previously-cherry-picked fix, and hopefully is a straightforward upgrade - let me know whether I should open a pull request for that, or if you'd prefer to apply the upgrade yourself.

**Comment by Neustradamus** [comment]
@jayaddison: Thanks for your comment :)

@xroche: There are 2 CVE fixes in latest Zlib 1.3.1 for minizip.
Have you received my emails?
I try to contact you in private.

**Pull request merged**